### PR TITLE
Issue 6843 - Add CI tests for logconv.py

### LINKED ISSUE #6843: Add CI tests for logconv.py
**Issue Description**
Add CI tests for logconv.py

**Package Version and Platform:**
 - Package and version: main branch back to 3.0

### PR DESCRIPTION
**Pull request by jchapma**
Description: logconv.py was added in 389-ds-base-3.0, we need some CI tests to help validate the upgrade of logconv.py to support JSON formatted access logs.

Fixes: https://github.com/389ds/389-ds-base/issues/6843

Reviewed by:

### CODE REVIEW

File: dirsrvtests/tests/suites/logging/logconv_test.py [hunk 1]
```diff
@@ -0,0 +1,992 @@
...
+        "unindexed_components": 0,
+        "operations": 2,
+        "results": 1
+    }
+    run_and_validate(expected, access_log, test_name="test_partially_unindexed_vlv_sort")
+
+def test_fully_unindexed_search(topology_st, get_access_log_path):
+    """Test that a fully unindexed search is correctly identified by logconv.
+
+    :id: 71f93a9f-3f9d-450e-ae98-79fabe5f21f4
```
Comment (droideck) [inline dirsrvtests/tests/suites/logging/logconv_test.py]: Duplicate UUID

File: dirsrvtests/tests/suites/logging/logconv_test.py [hunk 2]
```diff
@@ -0,0 +1,992 @@
...
+        2. StartTLS search completes without errors.
+        3. logconv returns correct updated stats.
+        4. Expected stats match logconv output.
+    """
+    access_log = truncate_access_log(get_access_log_path)
+
+    cmd = [
+        "ldapsearch",
+        "-x",
+        "-H", f"ldap://localhost:38901",
```
Comment (droideck) [inline dirsrvtests/tests/suites/logging/logconv_test.py]: I think the port shouldn't be hardcoded (in case, it's changed in the future)

File: dirsrvtests/tests/suites/logging/logconv_test.py [hunk 3]
```diff
@@ -0,0 +1,992 @@
...
+    inst = topology_st.standalone
+    access_log = truncate_access_log(get_access_log_path)
+
+    # Force an unindexed search
+    db_cfg = DatabaseConfig(inst)
+    default_idlistscanlimit = db_cfg.get_attr_vals_utf8('nsslapd-idlistscanlimit')
+    db_cfg.set([('nsslapd-idlistscanlimit', '100')])
+    users = UserAccounts(inst, DEFAULT_SUFFIX)
+    for i in range(101):
+        users.create_test_user(uid=i)
```
Comment (droideck) [inline dirsrvtests/tests/suites/logging/logconv_test.py]: I think it should be done with a fixture (and cleaned up accordingly)

File: dirsrvtests/tests/suites/logging/logconv_test.py [hunk 4]
```diff
@@ -0,0 +1,992 @@
...
+    raw_objects = DSLdapObjects(inst, basedn=DEFAULT_SUFFIX)
+    raw_objects.filter("(description=test*)")
+    db_cfg.set([('nsslapd-idlistscanlimit', default_idlistscanlimit)])
```
Comment (droideck) [inline dirsrvtests/tests/suites/logging/logconv_test.py]: It'll be safer to have it in `finally` block

**Review by droideck — CHANGES_REQUESTED** [review]
The CI fails (at least, I see it in BDB case). Probably, it's a bit flaky.

Additionally, I think we need some negative tests...
Something along these error conditions (at least to some basic degree):
- What happens when logconv.py fails?
- How are malformed log entries handled?
- Edge cases like empty or corrupted log files?

File: dirsrvtests/tests/suites/logging/logconv_test.py [hunk 5]
```diff
@@ -0,0 +1,980 @@
...
+        _test_ldapi,
+        _test_restart,
+        _test_invalid_filter,
+        _test_partially_unindexed_vlv_sort,
+        _test_fully_unindexed_search,
+        _test_ldaps,
+        _test_starttls,
+        _test_missing_access_log,
+        _test_invalid_access_log_format
+    ]
```
Comment (droideck) [inline dirsrvtests/tests/suites/logging/logconv_test.py]: what is the reason for such test suite structure?
It'll be harder to maintain, the output will be not in the expected pytest format, and it's hard to run one test without running all of them.

Pytest allows for test classes and will provide correct output:

So, I think, it should be something like this:

```python
# Each test should be independent
class TestLogConv:
    @pytest.mark.parametrize("log_format", ["default", "json"])
    def test_add_delete_operations(self, fresh_instance, log_format):
        # Test only add/delete
        
    @pytest.mark.parametrize("log_format", ["default", "json"])
    def test_modify_operations(self, fresh_instance, log_format):
        # Test only modify
```

Or just plainly without `class TestLogConv:`

File: dirsrvtests/tests/suites/logging/logconv_test.py [hunk 6]
```diff
@@ -0,0 +1,980 @@
...
+            time.sleep(poll_interval)
+
+@pytest.fixture
+def fresh_instance():
+    topo = create_topology({ReplicaRole.STANDALONE: 1})
+    inst = topo.standalone
+    inst.start()
+    yield inst
+    inst.stop()
+    inst.delete(escapehatch="i am sure")
```
Comment (droideck) [inline dirsrvtests/tests/suites/logging/logconv_test.py]: It needs a finalizer.

So, I think, it's better to reuse a usual topology approach:

```python


@pytest.fixture(scope="function")
def topology_st(request):
    """Create DS standalone instance"""

    topology = create_topology({ReplicaRole.STANDALONE: 1})

    def fin():
        if topology.standalone.exists():
            topology.standalone.delete()
    request.addfinalizer(fin)

    return topology

File: dirsrvtests/tests/suites/logging/logconv_test.py [hunk 7]
```diff
@@ -0,0 +1,980 @@
...
+            except (FileNotFoundError, RuntimeError) as e:
+                log.error(f"{test_name}: Error running logconv: {e}")
+                raise
+
+            summary = self.parse_summary(output)
+            for key, existing_val in summary.items():
+                if key in IGNORE_KEYS:
+                    continue
+                expected_val = expected.get(key, 0)
+                if existing_val < expected_val:
```
Comment (droideck) [inline dirsrvtests/tests/suites/logging/logconv_test.py]: What about `if existing_val > expected_val` case? In case of duplicates, or when we expect 0, or some other bug?

File: dirsrvtests/tests/suites/logging/logconv_test.py [hunk 8]
```diff
@@ -0,0 +1,980 @@
...
+        test_env.inst.config.set('nsslapd-ldapifilepath', f'/var/run/slapd-{test_env.inst.serverid}.socket')
+        test_env.inst.restart()
+        ldapi_socket = test_env.inst.config.get_attr_val_utf8('nsslapd-ldapifilepath').replace('/', '%2F')
+        cmd = [
+            "ldapsearch",
+            "-H", f"ldapi://{ldapi_socket}",
+            "-Y", "EXTERNAL",
+            "-b", "dc=example,dc=com"
+        ]
+        subprocess.run(cmd, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
```
Comment (droideck) [inline dirsrvtests/tests/suites/logging/logconv_test.py:679]: Not capturing output makes debugging failures difficult (in case, something will happen during the search - we can see the issue right away).

File: dirsrvtests/tests/suites/logging/logconv_test.py [hunk 9]
```diff
@@ -0,0 +1,980 @@
...
+
+def _test_add_del(test_env):
+
+    def _add_del_users(test_env):
+        users = UserAccounts(test_env.inst, DEFAULT_SUFFIX)
+        try:
+            for idx in range(3):
+                users.create_test_user(uid=str(idx), gid=str(idx))
+        except ldap.ALREADY_EXISTS as e:
+            log.warning(f"_test_add_del - User already exists: {e}")
```
Comment (droideck) [inline dirsrvtests/tests/suites/logging/logconv_test.py:250]: Do we really need it?
I think we expect the test to be isolated...
But it's a minor issue, I don't have a strong preference.

File: dirsrvtests/tests/suites/logging/logconv_test.py [hunk 10]
```diff
@@ -0,0 +1,980 @@
...
+        "fds_taken": 1,
+        "fds_returned": 1
+    }
+
+    test_env.truncate_logs()
+    test_env.run_operation(test_env.operation)
+
+def _test_ldapi(test_env):
+
+    def _ldapi_search(tetest_envt):
```
Comment (droideck) [inline dirsrvtests/tests/suites/logging/logconv_test.py]: **tetest_envt** typo:)

**Review by droideck — CHANGES_REQUESTED** [review]

File: dirsrvtests/tests/suites/logging/logconv_test.py (hunk 5 above)
Comment (jchapma) [reply dirsrvtests/tests/suites/logging/logconv_test.py]: > what is the reason for such test suite structure? It'll be harder to maintain, the output will be not in the expected pytest format, and it's hard to run one test without running all of them.
> 
> Pytest allows for test classes and will provide correct output:
> 
> So, I think, it should be something like this:
> 
> ```python
> # Each test should be independent
> class TestLogConv:
>     @pytest.mark.parametrize("log_format", ["default", "json"])
>     def test_add_delete_operations(self, fresh_instance, log_format):
>         # Test only add/delete
>         
>     @pytest.mark.parametrize("log_format", ["default", "json"])
>     def test_modify_operations(self, fresh_instance, log_format):
>         # Test only modify
> ```
> 
> Or just plainly without `class TestLogConv:`

I guess the main reason was because the sequence of each test was the same, truncate the logs, perform some operation, run logconv.py and compare the actual with the expected. I didn't like the idea of using a fixture for each logical step so I encapsulated all steps in a single class.  A single test_ function is used as a test runner to iterate over all the defined tests, which all have a defined structure, making it easy to extend,  well In my opinion. I did consider the option of defining each test separately but then I though when would we need to run a single test in isolation.

To be honest I didnt realise pytest has a  test class, let me look into that.

File: dirsrvtests/tests/suites/logging/logconv_test.py (hunk 6 above)
Comment (jchapma) [reply dirsrvtests/tests/suites/logging/logconv_test.py]: > It needs a finalizer.
> 
> So, I think, it's better to reuse a usual topology approach:
> 
> ```python
> @pytest.fixture(scope="function")
> def topology_st(request):
>     """Create DS standalone instance"""
> 
>     topology = create_topology({ReplicaRole.STANDALONE: 1})
> 
>     def fin():
>         if topology.standalone.exists():
>             topology.standalone.delete()
>     request.addfinalizer(fin)
> 
>     return topology
> ```

Yep, that is better

File: dirsrvtests/tests/suites/logging/logconv_test.py (hunk 7 above)
Comment (jchapma) [reply dirsrvtests/tests/suites/logging/logconv_test.py]: > What about `if existing_val > expected_val` case? In case of duplicates, or when we expect 0, or some other bug?

Initially the expected_val had to be == the expected_val for a test to pass, but with non test activity n the logs from plugins, etc made it very tricky. The best I could come up with is a form of "loose" validation where the actual matched or exceed the expected. Any suggestions on improving this ?

File: dirsrvtests/tests/suites/logging/logconv_test.py (hunk 8 above)
Comment (jchapma) [reply dirsrvtests/tests/suites/logging/logconv_test.py:679]: > Not capturing output makes debugging failures difficult (in case, something will happen during the search - we can see the issue right away).

Good point

File: dirsrvtests/tests/suites/logging/logconv_test.py (hunk 9 above)
Comment (jchapma) [reply dirsrvtests/tests/suites/logging/logconv_test.py:250]: > Do we really need it? I think we expect the test to be isolated... But it's a minor issue, I don't have a strong preference.

In a previous version all test for both formats were run on the same instance, so I added this to be sure. But in this version we do not.

File: dirsrvtests/tests/suites/logging/logconv_test.py (hunk 7 above)
Comment (droideck) [reply dirsrvtests/tests/suites/logging/logconv_test.py]: A few options I see:
- Categorize Keys by Validation Strategy (some strict, some more tolerant)
- Validate with a percentage tolerance for certain metrics
- Get baseline counts before the test and account for that

I'd say the final solution should be a combination of that.
This is the most important part of the test, so I think it's worth spending time calibrating.

**Comment by jchapma** [comment]
Hi @droideck, I think I have addressed all your concerns, if you get a chance could you re-review this PR. Thanks

**Comment by droideck** [comment]
> Hi @droideck, I think I have addressed all your concerns, if you get a chance could you re-review this PR. Thanks

Hi!
I noticed that some non-JSON tests are still failing:

```
FAILED dirsrvtests/tests/suites/logging/logconv_test.py::TestLogconv::test_starttls[default]
FAILED dirsrvtests/tests/suites/logging/logconv_test.py::TestLogconv::test_invalid_access_log_format[default]
FAILED dirsrvtests/tests/suites/logging/logconv_test.py::TestLogconv::test_ldaps[default]
```

Are these failures expected at this stage? I held off on following up as I thought you might still be investigating these issues.
Apologies for the lack of communication on my end...

**Comment by jchapma** [comment]
> > Hi @droideck, I think I have addressed all your concerns, if you get a chance could you re-review this PR. Thanks
> 
> Hi! I noticed that some non-JSON tests are still failing:
> 
> ```
> FAILED dirsrvtests/tests/suites/logging/logconv_test.py::TestLogconv::test_starttls[default]
> FAILED dirsrvtests/tests/suites/logging/logconv_test.py::TestLogconv::test_invalid_access_log_format[default]
> FAILED dirsrvtests/tests/suites/logging/logconv_test.py::TestLogconv::test_ldaps[default]
> ```
> 
> Are these failures expected at this stage? I held off on following up as I thought you might still be investigating these issues. Apologies for the lack of communication on my end...

Hey, ye these are expected as there is a bug in the original version. No worries. Thanks

File: dirsrvtests/tests/suites/logging/logconv_test.py [hunk 11]
```diff
@@ -0,0 +1,1094 @@
...
+            "operations": 1,
+            "paged_searches": 1,
+            "results": 1
+        }
+
+        output = self.run_logconv()
+        logconv_stats = self.extract_logconv_stats(output)
+        assert self.validate_logconv_stats(expected, logconv_stats, "test_paged_search")
+
+    def test_persistant_search(self):
```
Comment (droideck) [inline dirsrvtests/tests/suites/logging/logconv_test.py]:
```suggestion
    def test_persistent_search(self):
```

Comment (droideck) [reply dirsrvtests/tests/suites/logging/logconv_test.py]: And more typos like this in the test (doc string and the end of the test)

**Review by droideck — APPROVED** [review]
LGTM!
